Mathlib/NumberTheory/DirichletCharacter/Basic.lean

```diff
@@ -225,4 +224,35 @@ lemma primitiveCharacter_one (hn : n ≠ 0) :
       (isPrimitive_def _).1 (1 : DirichletCharacter R n).primitiveCharacter_isPrimitive,
       conductor_one hn]
 
+variable {S : Type} [CommRing S] {m : ℕ} (ψ : DirichletCharacter S m)
+
+/-- A Dirichlet character is odd if its value at -1 is -1. -/
+def Odd : Prop := ψ (-1) = -1
+
+/-- A Dirichlet character is even if its value at -1 is 1. -/
+def Even : Prop := ψ (-1) = 1
+
+lemma even_or_odd [NoZeroDivisors S] : ψ.Even ∨ ψ.Odd := by
+  suffices : ψ (-1) ^ 2 = 1
+  · convert sq_eq_one_iff.mp this
+  · rw [← map_pow _, neg_one_sq, map_one]
+
+lemma toUnitHom_odd_eval_neg_one (hψ : ψ.Odd) : ψ.toUnitHom (-1) = -1 := by
+  rw [← Units.eq_iff, MulChar.coe_toUnitHom]
+  exact hψ
+
+lemma toUnitHom_even_eval_neg_one (hψ : ψ.Even) : ψ.toUnitHom (-1) = 1 := by
+  rw [← Units.eq_iff, MulChar.coe_toUnitHom]
+  exact hψ
+
+lemma odd_eval_neg (x : ZMod m) (hψ : ψ.Odd) : ψ (- x) = - ψ x := by
+  rw [Odd] at hψ
+  rw [← neg_one_mul, map_mul]
+  simp [hψ]
+
+lemma even_eval_neg (x : ZMod m) (hψ : Even ψ) : ψ (- x) = ψ x := by
```

Comment:
Please rename these lemmas so that they can be used with dot notation. So
```suggestion
lemma Odd.toUnitHom_eval_neg_one (hψ : ψ.Odd) : ψ.toUnitHom (-1) = -1 := by
  rw [← Units.eq_iff, MulChar.coe_toUnitHom]
  exact hψ

lemma Even.toUnitHom_eval_neg_one (hψ : ψ.Even) : ψ.toUnitHom (-1) = 1 := by
  rw [← Units.eq_iff, MulChar.coe_toUnitHom]
  exact hψ

lemma Odd.eval_neg (x : ZMod m) (hψ : ψ.Odd) : ψ (- x) = - ψ x := by
  rw [Odd] at hψ
  rw [← neg_one_mul, map_mul]
  simp [hψ]

lemma Even.eval_neg (x : ZMod m) (hψ : ψ.Even) : ψ (- x) = ψ x := by
```